src/os/kstore/KStore.cc

```diff
@@ -2592,8 +2593,15 @@ void KStore::_dump_onode(OnodeRef o)
   }
 }
 
-void KStore::_do_read_stripe(OnodeRef o, uint64_t offset, bufferlist *pbl)
+void KStore::_do_read_stripe(OnodeRef o, uint64_t offset, bufferlist *pbl, bool cache_stripes)
 {
+  if(!cache_stripes) {
```

Comment:
i understand it's a pain to keep the read-once strips in the `onode_map`. but what if we perform read after write? then i think, we will be reading stale stripes after this change. am i missing anything?

Reply:
I think teh placement for this is wrong.. if there is something in o->pending_stripes, we should still read from it.  But if ``!cache_stripes``, we want to skip (just) the ``o->pending_stripes[offset] = *pbl;`` step below... right?

Reply:
> if there is something in o->pending_stripes, we should still read from it

yeah, agreed.

so, something like this:

```c++
  if (auto p = o->pending_stripes.find(offset);
      p != o->pending_stripes.end()) {
    *pbl = p->second;
  } else {
    string key;
    get_data_key(o->onode.nid, offset, &key);
    db->get(PREFIX_DATA, key, pbl);
    if (do_cache) {
      o->pending_stripes[offset] = *pbl;
    }
  }
```

in which i renamed `cache_stripes` to `do_cache`.

Reply:
> if there is something in o->pending_stripes, we should still read from it

`cache_stripes` will be false only in `KStore::read`, when perform write transactions, it will be true. On write process, we will try to read something from pending_stripes first, so we can still see our writes?

```
  if (auto p = o->pending_stripes.find(offset);
      p != o->pending_stripes.end()) {
    *pbl = p->second;
  } else {
    string key;
    get_data_key(o->onode.nid, offset, &key);
    db->get(PREFIX_DATA, key, pbl);
    if (do_cache) {
      o->pending_stripes[offset] = *pbl;
    }
  }
```
this way is actually right, but on a read-only process(who invoked `KStore::read`), we don't have to cache in-flight stripes since we would't read the same key twice.
am I right ?

Reply:
@Leeshine you are right. and i don't think you've covered the case of read-after-write. for instance, after your change, assuming we have following steps,

1. write "foo" to stripe 42. in that case, `cache_stripes` is true, so `pending_stripes[42] = "foo"`.
2. read "foo" from stripe 42. in that case, `cache_stripe` is false, so `_do_read_stripe()` will read directly from underlying `db`. so the stale data is returned.

Reply:
@tchaikov Hi, Kefu, do you mean that there are read-after-write OPs in **one** transaction? it seems OS transaction only contains write OPs, no read OP any ( https://github.com/ceph/ceph/blob/master/src/os/Transaction.h#L108 ).

Reply:
no. the two steps in my last comment are performed by two separated requests.

Reply:
After a transaction ends, the data is flushed. Subsequent reads will not read outdated data ？

Reply:
@liuchang0812 yeah, the pending_stripes is flushed in `KStore::_txc_finish()`. i think i was wrong.

the change looks good in general.